### PR TITLE
Support for external packages

### PR DESCRIPTION
**Pull request by pankajroark**
In this approach, we map a scattered truss into a shadow truss before any docker related operations. The shadow truss is updated when we observe any changes in the original truss, based on modification time.

Please refer to `docs/develop/external-packages.md` (in this PR) for details.

### CODE REVIEW

File: truss/contexts/local_loader/truss_module_loader.py [hunk 1]
```diff
@@ -106,12 +124,21 @@ def _unload_truss_modules(
     truss_dir: str,
     truss_module_name: str,
     bundled_packages_dir_name: str = None,
+    external_packages_dirs: List[str] = None,
 ):
     modules_to_unload = [truss_module_name]
+
+    def _add_relative_dir_to_unload(dir_nae: str):
```
Comment (bolasm) [inline truss/contexts/local_loader/truss_module_loader.py:131]: nit: s/dir_nae/dir_name/g

File: truss/contexts/local_loader/truss_module_loader.py [hunk 2]
```diff
@@ -106,12 +124,21 @@ def _unload_truss_modules(
...
+        path = Path(truss_dir) / dir_nae
+        if path.exists():
+            modules_to_unload.extend(_sub_dirnames(path))
```
Comment (bolasm) [inline truss/contexts/local_loader/truss_module_loader.py:134]: Should we warn if the path can't be resolved?

File: truss/truss_gatherer.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    packages_dir_path_in_shadow = (
+        shadow_truss_path / handle.spec.config.bundled_packages_dir
+    )
+    packages_dir_path_in_shadow.mkdir(exist_ok=True)
+    for path in handle.spec.external_packages_dirs_paths:
+        if not path.is_dir():
+            raise ValueError(
+                f"External packages directory at {path} is not a directory"
+            )
+        for sub_path in path.iterdir():
```
Comment (bolasm) [inline truss/truss_gatherer.py:41]: Can you add a comment here about not copying the directory directly to mimic the path import of the local loader?

File: truss/truss_handle.py [hunk 4]
```diff
@@ -670,6 +699,45 @@ def calc_patch(self, prev_truss_hash: str) -> Optional[PatchDetails]:
...
+        if self.no_external_packages():
+            return max_mod_time
+
+        for path in self.spec.external_packages_dirs_paths:
+            max_mod_time_for_path = get_max_modified_time_of_dir(path)
+            if max_mod_time_for_path > max_mod_time:
+                max_mod_time = max_mod_time_for_path
+        return max_mod_time
+
+    def no_external_packages(self) -> bool:
```
Comment (bolasm) [inline truss/truss_handle.py:731]: should this be a property?

File: truss/truss_handle.py [hunk 5]
```diff
@@ -670,6 +699,45 @@ def calc_patch(self, prev_truss_hash: str) -> Optional[PatchDetails]:
...
+        are caches and resused.
+        """
+        from truss.truss_gatherer import gather
+
+        if not self.is_scattered():
+            return self._truss_dir
+
+        return gather(self._truss_dir)
+
+    def max_modified_time(self) -> float:
```
Comment (bolasm) [inline truss/truss_handle.py:719]: should this be a property?

**Review by bolasm — APPROVED** [review]
lgtm. Just a few comments/questions. I'll close out my other PR

File: truss/contexts/local_loader/truss_module_loader.py (hunk 2 above)
Comment (pankajroark) [reply truss/contexts/local_loader/truss_module_loader.py:134]: I think we should validate this at the TrussHandle creation to consolidate the check at one place. Let me do that.

**Pull request merged**